### PR TITLE
[#2862] Modified docstring for apply_unitary()

### LINKED ISSUE #2862: Document in cirq.apply_unitary's docstring that it can operate on qudit subspaces
If you have a state tensor defined over qutrits, and you call `cirq.apply_unitary` targeting this state tensor but give it a qubit gate, apply_unitary will operate on the 01 subspace. This isn't currently documented but it should be mentioned in the docstring.

### PR DESCRIPTION
**Pull request by aditya-giri**
This fixes #2862 
@Strilanc @vtomole - Since I'm new to Cirq's notation, please do go through this and let me know if it makes sense.

### CODE REVIEW

**Comment by aditya-giri** [comment]
@googlebot I signed it!

**Comment by aditya-giri** [comment]
@vtomole bump ^

File: cirq/protocols/apply_unitary_protocol.py [hunk 1]
```diff
@@ -323,6 +323,11 @@ def apply_unitary(unitary_value: Any,
         all three of these cases. In all cases `args.target_tensor` and
         `args.available_buffer` may have been mutated.
 
+        Note: If `args` specifies a tensor defined over d-level quantum systems
+        and `unitary_value` operates on k-level quantum systems (k < d), then
+        the effect of `unitary_value` will be restricted to the subspace
+        consisting of the first `k` levels of the system.
+
```
Comment (dabacon) [inline cirq/protocols/apply_unitary_protocol.py:330]: I believe this also works in the case that there are qudits of different shapes.  Here is an example
```
target = np.zeros(6)
target[1] = 1
target.shape=(3, 2)
buffer = np.zeros_like(v)
out = cirq.apply_unitary(cirq.X, cirq.ApplyUnitaryArgs(v, available_buffer=buffer, axes=[0]))
print(f'input: {target}')
print(f'output: {out}')
```
which prints
```
input: [[0. 1.]
 [0. 0.]
 [0. 0.]]
output: [[0. 0.]
 [0. 1.]
 [0. 0.]]
```

Could you reword this for this more general case?  Something like "tensor defined with shape (d_1, d_2, ...d_n)" and such?

**Comment by aditya-giri** [comment]
Sure, let me do that.

**Comment by balopat** [comment]
Friendly ping @aditya-giri - can you commit the rewording so we can merge this PR? Thank you!

**Comment by aditya-giri** [comment]
Oops, this skipped my mind completely. Sorry 'bout that. Give me roughly 24 hours and I'll make the change :) 
@dabacon just want to confirm that in the changes that you suggested, `v` should be replaced by `target`...

File: cirq/protocols/apply_unitary_protocol.py (hunk 1 above)
Comment (aditya-giri) [reply cirq/protocols/apply_unitary_protocol.py:330]: [The docstring for ApplyUnitaryArgs](https://github.com/quantumlib/Cirq/blob/a344fb715995b00fa87bf9d2bef8aab04d076854/cirq/protocols/apply_unitary_protocol.py#L64) seems to indicate that `target_tensor` should have the shape `(2,2,2,...,2)`. 
Should that be changed as well?

**Comment by aditya-giri** [comment]
^ bump, could one of you clarify please? :)

File: cirq/protocols/apply_unitary_protocol.py (hunk 1 above)
Comment (balopat) [reply cirq/protocols/apply_unitary_protocol.py:330]: Yes, that should be modified as well AFAICT, I'm going to review this - but I'll need some time to parse this code, I'm new to it. Maybe @cduck can jump in too?

Comment (aditya-giri) [reply cirq/protocols/apply_unitary_protocol.py:330]: Yep, that would be good. I just need a little more clarity on this.

Comment (aditya-giri) [reply cirq/protocols/apply_unitary_protocol.py:330]: Friendly ping to remind you guys about this :)

Comment (balopat) [reply cirq/protocols/apply_unitary_protocol.py:330]: Okay, I spent some time understanding the code. Here's the recommendation: 
- Yes, the docstring should be modified for ApplyUnitaryArgs as well, arbitrary tensors are accepted 
- Yes, please update to the generic (d_1, d_2, ..., d_n) shape in both places.

If you want a more complex working example try to parse through this:

```
state = np.arange(3 * 2 * 4,dtype=np.complex64).reshape((3,2,4))
result = cirq.X._apply_unitary_(cirq.ApplyUnitaryArgs(state, np.zeros_like(state), [2]))
print(state)
print(result)
```

Prints: 

```
# (a non-normalized) state - a qutrit (0th axis), qubit (1st axis) and a ququart (2nd axis)

[[[ 0.+0.j  1.+0.j  2.+0.j  3.+0.j]
  [ 4.+0.j  5.+0.j  6.+0.j  7.+0.j]]

 [[ 8.+0.j  9.+0.j 10.+0.j 11.+0.j]
  [12.+0.j 13.+0.j 14.+0.j 15.+0.j]]

 [[16.+0.j 17.+0.j 18.+0.j 19.+0.j]
  [20.+0.j 21.+0.j 22.+0.j 23.+0.j]]]

# result - see the cols 0th and 1st "cols" swapped 
# which is the 0th and 1st subspace of the 2nd axis (a ququart)

[[[ 1.+0.j  0.+0.j  0.+0.j  0.+0.j]
  [ 5.+0.j  4.+0.j  0.+0.j  0.+0.j]]

 [[ 9.+0.j  8.+0.j  0.+0.j  0.+0.j]
  [13.+0.j 12.+0.j  0.+0.j  0.+0.j]]

 [[17.+0.j 16.+0.j  0.+0.j  0.+0.j]
  [21.+0.j 20.+0.j  0.+0.j  0.+0.j]]]

```

**Comment by balopat** [comment]
> ^ bump, could one of you clarify please? :)

Clarified in the above comments @aditya-giri - hope this helps, thank you again for your patience on this.

**Comment by aditya-giri** [comment]
thank you @balopat, I will update the PR asap

**Comment by balopat** [comment]
> thank you @balopat, I will update the PR asap

Friendly ping @aditya-giri - do you have an update on this? Thank you!

**Comment by balopat** [comment]
Hi @aditya-giri - I'm going to close this now, please feel free to reopen when you have time to write it. In the meantime I'll unassign you from the issue. Thanks!

**Pull request closed without merging**